airbyte-integrations/connectors/source-orb/source_orb/source.py

```diff
@@ -177,6 +180,257 @@ def transform_record(self, subscription_record):
         return subscription_record
 
 
+# helpers for working with pendulum dates and strings
+def to_datetime(time) -> pendulum.DateTime:
+    if time is None:
+        return None
+    elif isinstance(time, pendulum.DateTime):
+        return time
+    elif isinstance(time, str):
+        return pendulum.parse(time)
+    else:
+        raise TypeError(f"Cannot convert input of type {type(time)} to DateTime")
+
+
+def to_utc_isoformat(time) -> str:
+    if time is None:
+        return None
+    elif isinstance(time, pendulum.DateTime):
+        return time.in_timezone("UTC").isoformat()
+    elif isinstance(time, str):
+        return pendulum.parse(time).in_timezone("UTC").isoformat()
+    else:
+        raise TypeError(f"Cannot convert input of type {type(time)} to isoformat")
+
+
+def chunk_date_range(start_date: pendulum.DateTime, end_date: Optional[pendulum.DateTime] = None) -> Iterable[pendulum.Period]:
+    """
+    Yields a list of the beginning and ending timestamps of each day between the start date and now.
+    The return value is a pendulum.period
+    """
+    one_day = pendulum.duration(days=1)
+    end_date = end_date or pendulum.now()
+
+    # Each stream_slice contains the beginning and ending timestamp for a 24 hour period
+    chunk_start_date = start_date
+    while chunk_start_date < end_date:
+        chunk_end_date = min(chunk_start_date + one_day, end_date)
+        yield pendulum.period(chunk_start_date, chunk_end_date)
+        chunk_start_date = chunk_end_date
+
+
+class SubscriptionUsage(IncrementalOrbStream):
+    """
+    API Docs: https://docs.withorb.com/docs/orb-docs/api-reference/operations/get-a-subscription-usage
+    """
+
+    cursor_field = "timeframe_start"
+
+    def __init__(
+        self,
+        subscription_usage_grouping_key: Optional[str] = None,
+        plan_id: Optional[str] = None,
+        end_date: Optional[pendulum.DateTime] = None,
+        **kwargs,
+    ):
+
+        self.subscription_usage_grouping_key = subscription_usage_grouping_key
+        self.plan_id = plan_id
+        self.end_date = end_date
+        super().__init__(**kwargs)
+
+    @property
+    def primary_key(self) -> Iterable[str]:
+        key = ["subscription_id", "billable_metric_id", "timeframe_start"]
+
+        # If a grouping key is present, it should be included in the primary key
+        if self.subscription_usage_grouping_key:
+            key.append(self.subscription_usage_grouping_key)
+
+        return key
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        # use a regex capture group to extract the subscription ID from the request URL
+        # so that we can include it in the returned record
+        subscription_id = re.search("/subscriptions/(.*)/usage", response.request.url).group(1)
```

Comment:
you should be able to get the `subscription_id` from the `stream_slice` if you replace the `**kwargs` argument with the [same signature](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http.py#L203-L210) as the `HttpStream` (the parent class)

(you could also get the `stream_slice` from `kwargs['stream_slice']` but **kwargs are tricky because they obfuscate the interface

Reply:
Ahh perfect, I was looking for a way to access the stream_slice in parse_response but was confused after using the same signature as the IncrementalOrbStream class and didn't think to check the HttpStream class! This is much cleaner